File: src/main/java/io/polymorphicpanda/panda/ecs/util/identity/RecyclingIdentityFactory.java
package io.polymorphicpanda.panda.ecs.util.identity;

import java.util.BitSet;

import io.polymorphicpanda.panda.ecs.util.collection.IntDeque;

/**
 * @author Ranie Jade Ramiso
 */
public class RecyclingIdentityFactory extends ForwardingIdentityFactory {
    private final BitSet recycled = new BitSet();
    private final IntDeque limbo = new IntDeque();

    public RecyclingIdentityFactory(IdentityFactory delegate) {
        super(delegate);
    }

    @Override
    public int generate() {
        if (!limbo.isEmpty()) {
            int id = limbo.poll();
            recycled.set(id, false);
            return id;
        }
        return super.generate();
    }

    public void free(int id) {
        limbo.insert(id);
        recycled.set(id);
    }
}

File: src/main/kotlin/org/pandaframework/ecs/util/identity/RecyclingIdentityFactory.kt
package org.pandaframework.ecs.util.identity

import java.util.*

class RecyclingIdentityFactory (delegate: IdentityFactory): ForwardingIdentityFactory(delegate) {
    private val recycled: BitSet = BitSet()
    private val limbo: MutableList<Int> = mutableListOf()

    override fun generate(): Int {
        if (limbo.isEmpty()) {
            return super.generate()
        }
        val id = limbo.removeAt(0)
        recycled.set(id, false)
        return id
    }

    fun free(identity: Int) {
        limbo.add(identity)
        recycled.set(identity)
    }
}
